### DISCUSSION TITLE
Directory structure when contributing a custom model implementation

### DISCUSSION THREAD

**Discussion by ssiegel95** [kubeflow/kfserving#3807] [Q&A] [answered]
Hi, I'm part of team that's planning to propose and implement a new serving runtime for timeseries models. At present the plan is to use the custom runtime template in a similar way that [custom_model](https://github.com/kserve/kserve/tree/master/python/custom_model) is implemented. We noticed that this has a flat directory structure as opposed to, say, something like the [huggingfaceserver](https://github.com/kserve/kserve/tree/master/python/huggingfaceserver). Is there a particular reason why these two implementations are organized differently? We'd like to conform to kserve best practices. Thanks.

**Comment by terrytangyuan** [comment, marked as answer]
I don't think the folder structure matters. What matters is that the model servers are compatible with KServe's inference protocols and that there are Dockerfiles associated with them at the parent `/python` folder, e.g . [custom_model.Dockerfile](https://github.com/kserve/kserve/blob/master/python/custom_model.Dockerfile) and [huggingface_server.Dockerfile](https://github.com/kserve/kserve/blob/master/python/huggingface_server.Dockerfile). 

The HuggingFace model server is higher-level wrapper with underlying engines like vLLM. The vLLM engine is currently in a separate `huggingfaceserver/vllm/` folder but it could also live in the same directory as huggingfaceserver, just like normal Python package structure. As long as your Dockerfile knows how to find the source code (more specifically the entrypoint of the Python module in that directory) of the model server, it should be fine.

**Reply by ssiegel95** [reply]
Perfect, thank you @terrytangyuan.
